File: python/528. Random Pick with Weight.py
class Solution(object):
    
    def __init__(self, w):
        """
        :type w: List[int]
        """
        self.ws = []
        s = 0
        for x in w:
            s += x
            self.ws.append(s)
        self.s = s

    def pickIndex(self):
        """
        :rtype: int
        """
        w = random.randint(1, self.s)
        return bisect.bisect_left(self.ws, w)


# Your Solution object will be instantiated and called as such:
# obj = Solution(w)
# param_1 = obj.pickIndex()

File: go/src/leetcode/528. Random Pick with Weight /528. Random Pick with Weight.go
package main

import (
	"fmt"
	"math/rand"
)

type Solution struct {
	ws []int
	s int
}


func Constructor(w []int) Solution {
	ws := []int{}
	var s int
	for _, t := range w {
		s += t
		ws = append(ws, s)
	}
	return Solution{
		ws: ws,
		s: s,
	}
}


func (this *Solution) PickIndex() int {
	target := rand.Int31n(int32(this.s)) + 1
	i := lowerBound(this.ws, int(target))
	fmt.Println(i)
	return i
}

func lowerBound(ws []int, target int) int {
	l, r := 0, len(ws)
	var m int
	for l < r {
		m = l + (r-l)>>1
		if ws[m] >= target {
			r = m
		} else {
			l = m + 1
		}
	}
	return l
}

func main() {
	obj := Constructor([]int{1,3})
	obj.PickIndex()

}


/**
 * Your Solution object will be instantiated and called as such:
 * obj := Constructor(w);
 * param_1 := obj.PickIndex();
 */
